In [3]:
import pandas as pd


In [4]:
import numpy as np


In [5]:
tr = pd.read_table('.//drugsComTrain_raw.tsv')

In [6]:
tr.head()

Unnamed: 0                  drugName                     condition  \
0      206461                 Valsartan  Left Ventricular Dysfunction   
1       95260                Guanfacine                          ADHD   
2       92703                    Lybrel                 Birth Control   
3      138000                Ortho Evra                 Birth Control   
4       35696  Buprenorphine / naloxone             Opiate Dependence   

                                              review  rating  \
0  "It has no side effect, I take it in combinati...     9.0   
1  "My son is halfway through his fourth week of ...     8.0   
2  "I used to take another oral contraceptive, wh...     5.0   
3  "This is my first time using any form of birth...     8.0   
4  "Suboxone has completely turned my life around...     9.0   

                date  usefulCount  
0       May 20, 2012           27  
1     April 27, 2010          192  
2  December 14, 2009           17  
3   November 3, 2015           10  
4  November 27, 2016           37

There are five potential independent variables, one target variable, and a unique id column, for total of 7 columns.

In [7]:
tr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 161297 entries, 0 to 161296
Data columns (total 7 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Unnamed: 0   161297 non-null  int64  
 1   drugName     161297 non-null  object 
 2   condition    160398 non-null  object 
 3   review       161297 non-null  object 
 4   rating       161297 non-null  float64
 5   date         161297 non-null  object 
 6   usefulCount  161297 non-null  int64  
dtypes: float64(1), int64(2), object(4)
memory usage: 8.6+ MB


There are very few missing values, only 'condition' has missing values of about 1,000, less than 1% of total sample.

tr.drugName.sort_values()

In [11]:
len(tr.condition.unique())

885

In [12]:


count = 0
sets = {}
for i in tr.condition:
    if sets.get(i, 0) >0:
        sets[i]+=1
    else:
        sets[i]=1


In [13]:
len(sets.values())

885

In [34]:
sets

{'Left Ventricular Dysfunction': 19,
 'ADHD': 3383,
 'Birth Control': 28788,
 'Opiate Dependence': 1079,
 'Benign Prostatic Hyperplasia': 561,
 'Emergency Contraception': 2463,
 'Bipolar Disorde': 4224,
 'Epilepsy': 358,
 'Migraine Prevention': 1413,
 'Depression': 9069,
 "Crohn's Disease": 168,
 'Cough': 927,
 'Obesity': 3568,
 'Urinary Tract Infection': 1316,
 'ibromyalgia': 1791,
 'Chronic Myelogenous Leukemia': 89,
 'HIV Infection': 615,
 'Insomnia': 3673,
 'Rheumatoid Arthritis': 995,
 'Vaginal Yeast Infection': 2274,
 'Chlamydia Infection': 273,
 'Hirsutism': 74,
 'Panic Disorde': 1463,
 'Migraine': 1694,
 nan: 899,
 'Pain': 6145,
 'Irritable Bowel Syndrome': 1014,
 'Osteoarthritis': 1239,
 'Constipation': 1595,
 'Bowel Preparation': 1859,
 'Psychosis': 73,
 'Muscle Spasm': 1244,
 'Hepatitis C': 656,
 'Overactive Bladde': 683,
 'Diabetes, Type 2': 2554,
 'Asthma, Maintenance': 501,
 'Non-Small Cell Lung Cance': 94,
 'Schizophrenia': 594,
 'Dysuria': 73,
 'Smoking Cessation': 1780

It initially appears there are 885 unique condtions, but some entries are errors (i.e. "users found...")overlook for now.


tr["condition_Cl"] = tr["condition"].map(lamdba x for x if "users" not in x else null)

df[condition_Cl] = [null for condition in tr.condition if "user" in condition]

In [18]:


count = 0
sets = {}
for i in tr.drugName:
    if sets.get(i, 0) >0:
        sets[i]+=1
    else:
        sets[i]=1


In [19]:
len(sets)

3436

In [22]:
import matplotlib.pyplot as plt

In [25]:
sets

{'Valsartan': 69,
 'Guanfacine': 127,
 'Lybrel': 10,
 'Ortho Evra': 219,
 'Buprenorphine / naloxone': 437,
 'Cialis': 279,
 'Levonorgestrel': 3657,
 'Aripiprazole': 500,
 'Keppra': 117,
 'Ethinyl estradiol / levonorgestrel': 1888,
 'Topiramate': 606,
 'L-methylfolate': 73,
 'Pentasa': 37,
 'Dextromethorphan': 122,
 'Nexplanon': 2156,
 'Liraglutide': 755,
 'Trimethoprim': 8,
 'Amitriptyline': 695,
 'Lamotrigine': 591,
 'Nilotinib': 17,
 'Atripla': 68,
 'Trazodone': 584,
 'Etonogestrel': 3336,
 'Etanercept': 152,
 'Tioconazole': 706,
 'Azithromycin': 548,
 'Eflornithine': 31,
 'Daytrana': 45,
 'Ativan': 220,
 'Imitrex': 114,
 'Sertraline': 1360,
 'Toradol': 117,
 'Viberzi': 90,
 'Mobic': 84,
 'Dulcolax': 396,
 'Morphine': 221,
 'MoviPrep': 147,
 'Trilafon': 4,
 'Fluconazole': 200,
 'Contrave': 920,
 'Clonazepam': 831,
 'Metaxalone': 75,
 'Venlafaxine': 1016,
 'Ledipasvir / sofosbuvir': 195,
 'Symbyax': 37,
 'Tamsulosin': 181,
 'Doxycycline': 627,
 'Dulaglutide': 231,
 'Intuniv': 84,
 'Bu

The data is fairly balanced by drug.

In [13]:
len(set(tr.drugName.sort_values()))

3436

It initially appears there are 3436 unique condtions, but some entries are errors (i.e. "users found...")overlook for now.

In [9]:
sm=0
for i in range(10):
    sm += len(tr['review'][i])


sm/11/5

90.0

About 90 words per review, given sample of 11 and average 5 words per sentence.

Example below:


In [22]:
tr.review[22]

'"Nexplanon does its job. I can have worry free sex. The only thing is that my periods are sometimes light and sometimes heavy. Sometimes they go away and sometimes they show up unexpected. I also feel somewhat depressed. Not sure if its Nexplanon or not. I&#039;ve had Nexplanont for about 2 months now, but despite the side effects its the most effective birth control I&#039;ve ever used and I do not plan on taking it out."'

In [40]:
#from nltk import word_tokenize
from nltk.tokenize import RegexpTokenizer

basic_token_pattern = r"(?u)\b\w\w+\b"

tokenizer = RegexpTokenizer(basic_token_pattern)


In [46]:
from nltk import FreqDist

expand below so all rowas are indivually tokenized

In [27]:
ttest= tokenizer.tokenize(tr.review[22])

In [28]:
FreqDist(ttest)

FreqDist({'its': 3, 'sometimes': 3, 'and': 3, 'Nexplanon': 2, 'they': 2, 'not': 2, '039': 2, 've': 2, 'the': 2, 'does': 1, ...})

Expand below to include all rows.

In [31]:
sample = tr[0:1000]

In [38]:
com = ""
for i in range(1000):
    com += sample['review'][i]

In [52]:
com = com.lower()

In [53]:
ttestw= tokenizer.tokenize(com)

fw = FreqDist(ttestw)

fw

In [42]:
from nltk.corpus import stopwords


In [54]:

stopwords_list = stopwords.words('english')
#stopwords_list += list(string.punctuation)
#stopwords_list += ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']

w_words_stopped = [word for word in ttestw if word not in stopwords_list]

In [55]:
FreqDist(w_words_stopped)

FreqDist({'039': 1542, 'side': 341, 'day': 341, 'taking': 326, 'years': 308, 'pain': 299, 'started': 293, 'effects': 293, 'days': 287, 'first': 287, ...})

In [56]:
w_words_stoppedC = [word for word in w_words_stopped if '039' not in word]

In [57]:
FreqDist(w_words_stoppedC)

FreqDist({'side': 341, 'day': 341, 'taking': 326, 'years': 308, 'pain': 299, 'started': 293, 'effects': 293, 'days': 287, 'first': 287, 'get': 267, ...})